src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/DecryptorPalOpenSsl.Decode.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Security.Cryptography;
+using System.Security.Cryptography.Pkcs;
+using System.Security.Cryptography.X509Certificates;
+
+using Microsoft.Win32.SafeHandles;
+
+namespace Internal.Cryptography.Pal.OpenSsl
+{
+    internal sealed partial class DecryptorPalOpenSsl : DecryptorPal
+    {
+        public static DecryptorPalOpenSsl Decode(
+            byte[] encodedMessage,
+            out int version,
+            out ContentInfo contentInfo,
+            out AlgorithmIdentifier contentEncryptionAlgorithm,
+            out X509Certificate2Collection originatorCerts,
+            out CryptographicAttributeObjectCollection unprotectedAttributes)
+        {
+            SafeCmsHandle cmsHandle = Interop.Crypto.CmsDecode(encodedMessage, encodedMessage.Length);
+            Interop.Crypto.CheckValidOpenSslHandle(cmsHandle);
+
+            string oid = cmsHandle.GetContentType();
+
+            if (oid != Oids.Pkcs7Enveloped)
+                throw new CryptographicException(SR.Cryptography_Cms_Invalid_Message_Type);
+
+            originatorCerts = cmsHandle.GetOriginatorCerts();
+            contentInfo = cmsHandle.GetEmbeddedContent();
+            RecipientInfoCollection recipientInfos = cmsHandle.GetRecipients();
+
+            try
+            {
+                // Some fields of the CMS are not exposed by the OpenSSL CMS API, we have to parse this values from
+                // the DER encoded message. In case there's an error decoding we need to throw an exception and release the
+                // message and certificates as at this point it makes no sense to assume metadata is correct nor that decrypt
+                // can be called after this.
+                ParseMissingValues(encodedMessage, out version, out contentEncryptionAlgorithm, out unprotectedAttributes);
+            }
+            catch (InvalidOperationException e)
+            {
+                foreach (X509Certificate2 cert in originatorCerts)
+                    cert.Dispose();
```

Comment:
Please use explicit braces.  I know we sometimes skip it on an if; but it feels wrong to skip them on a foreach :).

Consider nulling out the originatorCerts collection after loop-Disposing; just to avoid sending a bunch of disposed things out into the world.